File: JULY-2025/29-07-2025/p2411.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer[]
     */
    function smallestSubarrays($nums) {
         $length = count($nums);
    $lastSeen = array_fill(0, 30, 0);
    $result = array_fill(0, $length, 1);

    for ($i = $length - 1; $i >= 0; $i--) {
        for ($bit = 0; $bit < 30; $bit++) {
            if (($nums[$i] & (1 << $bit)) > 0) {
                $lastSeen[$bit] = $i;
            }
            $result[$i] = max($result[$i], $lastSeen[$bit] - $i + 1);
        }
    }

    return $result;
    }
}

File: JULY-2025/29-07-2025/p2411.rb
# @param {Integer[]} nums
# @return {Integer[]}
def smallest_subarrays(nums)
  n = nums.length

  # Calculate the smallest subarray length for each position
  result = Array.new(n, 1)
  last_seen = Array.new(32, -1)

  (n - 1).downto(0) do |i|
    nums[i].bit_length.times do |bit|
      last_seen[bit] = i if nums[i][bit] == 1
    end

    max_index = i
    last_seen.each do |index|
      max_index = [max_index, index].max if index != -1
    end

    result[i] = max_index - i + 1
  end

  result
end